Get-Bqjob + tests

ahandley opened the pull request:
First cmdlet for job functionality.  Does Lists and Gets.

Same general template as the other 3 sections.

iantalarico commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -0,0 +1,132 @@
+    /// <para type="description">
+    /// If no Job is specified though ID parameter or object via pipeline, this cmdlet lists all jobs that 
+    /// you started in the specified project. If a Job is specified, it will return a descriptor object 
+    /// for that job. Listing requires Can View or Is Owner project roles. Viewing information about a 
```
Should these be "View" and "Owner"? Can View and Is Owner seem odd

ahandley replied:
I took those permissions directly off of the BQ Jobs API page.  I do see that they're Legacy permissions and that the IAM roles are the way to go.  Should I keep these roles to be consistent with the docs or should I put the new IAM roles?  

(I sent feedback on the page mentioning this as well.  Is that read at all, or should I file a bug on the page for being out of date?)

iantalarico commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -0,0 +1,132 @@
+    /// you started in the specified project. If a Job is specified, it will return a descriptor object 
+    /// for that job. Listing requires Can View or Is Owner project roles. Viewing information about a 
+    /// specific job requires that you’re the one who ran the job, or that you have the Is Owner project 
+    /// role. Job information is stored for a six month period after its creation.
```
Job information is stored for a six month period after its creation.

Maybe:

Job information is stored for six months after its creation.

iantalarico commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -0,0 +1,132 @@
+    /// </para>
+    /// <example>
+    ///   <code>PS C:\> Get-BqJob</code>
+    ///   <para>This will list all past or present jobs from the default project.</para>
```
I think you can drop "This will" and say "Lists" in all of these

iantalarico commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -0,0 +1,132 @@
+            }
+        }
+
+        public void DoListRequest(JobsResource.ListRequest request)
```
comments on functions?

ahandley replied:
I figured the names would be self-documenting, but I threw on some additional details.

iantalarico commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -0,0 +1,132 @@
+        {
+            do
+            {
+                JobList response = request.Execute();
```
Catch GoogleApiExceptions?

ahandley replied:
got it.

iantalarico commented on Google.PowerShell.IntegrationTests/BigQuery/BigQuery.BqJob.Tests.ps1:
```diff
@@ -0,0 +1,55 @@
+            Add-BqTabledata $filename CSV -SkipLeadingRows 1
+    }
+
+    It "should list jobs from the past 6 months" {
```
Is this really testing anything we have control over?

iantalarico commented on Google.PowerShell.IntegrationTests/BigQuery/BigQuery.BqJob.Tests.ps1:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    It "should handle projects that the user does not have permissions for" {
+        { Get-BqJob -Project $accessErrProject -ErrorAction Stop } | Should Throw 400
```
Should 400 be in quotes?  If nothing else to be consistent?

ahandley replied:
PS auto-boxes/types, but yes.  For consistency, it should.

ahandley commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -0,0 +1,132 @@
+    /// Lists all jobs that you started in the specified project or returns information about a specific job.
+    /// </para>
+    /// <para type="description">
+    /// If no Job is specified though ID parameter or object via pipeline, this cmdlet lists all jobs that 
```
Fixed that paragraph entirely.  A lot of it was language directly from the API page, so it was stitched together somewhat awkwardly.

ahandley commented:
Fixed a bunch of stuff.  PTAL.

iantalarico commented on Google.PowerShell.IntegrationTests/BigQuery/BigQuery.BqJob.Tests.ps1:
```diff
@@ -14,11 +14,12 @@ Describe "Get-BqJob" {
     It "should list jobs from the past 6 months" {
         $jobs = Get-BqJob
-        $jobs.Count | Should BeGreaterThan 0
+        $jobs.Count | Should BeGreaterThan 1
```
This just tests that a job exists.  Is this something we need a test for?  We test the get calls below

ahandley replied:
Yes, because that tests the "list" functionality of the cmdlet.  It checks that we can access job history (and presently running jobs if there are any), whereas the other two test getting a specific job.

iantalarico replied:
SGTM

iantalarico commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -15,19 +16,18 @@ namespace Google.PowerShell.BigQuery
-    /// for that job. Listing requires Can View or Is Owner project roles. Viewing information about a 
-    /// specific job requires that you’re the one who ran the job, or that you have the Is Owner project 
-    /// role. Job information is stored for a six month period after its creation.
+    /// If no Job is specified through the JobId parameter or object via pipeline, a list of all jobs in the 
```
Very nice :)

ILMTitan commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -0,0 +1,148 @@
+        {
+            if (JobId == null && InputObject == null)
+            {
+                DoListRequest(Service.Jobs.List(Project));
```
It is a little odd that your are passing a request here. The standard structure for our Cmdlets has been to build the request inside the specific function, and then have that function return either the single resource or an IEnumerable of resources, and do the WriteObject in the ProcessRecord. This separates the Google Api stuff (the request, response etc) and the PowerShell Cmdlet stuff (ProcessRecord override, WriteObject).

ahandley replied:
I refactored one, and sort of refactored the other.  I couldn't find a more elegant way of doing it other than bringing both of the catch blocks into ProcessRecord, which severely clutters it.  So I'm not passing in a request anymore, but the helper function also doesn't return anything.  I think it's a decent compromise, while keeping the code readable.

ILMTitan commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -0,0 +1,148 @@
+        /// Executes a List Jobs request and writes returned objects or errors.
+        /// </summary>
+        /// <param name="request">Pre-built request ready to be executed</param>
+        public void DoListRequest(JobsResource.ListRequest request)
```
There appears to be a stateFilter in the Jobs.List API. As it is optional, it is probably a property rather than a parameter. There should be a Cmdlet parameter that sets it.
There is also an allUsers boolean. There should be a switch parameter for that as well.

ahandley replied:
Added both of these.  Tests will come later when I have Start- and Stop- usable.

ILMTitan commented on Google.PowerShell.IntegrationTests/BigQuery/BigQuery.BqJob.Tests.ps1:
```diff
@@ -0,0 +1,58 @@
+        $return = Get-BqJob -JobId $job.JobReference.JobId
+        $return.JobReference.JobId | Should Be $job.JobReference.JobId
+    }
+
```
Add a test like:

$jobs = Get-BqJob -Project $someOtherProject
$job = $jobs[0]
$return = $job | Get-BqJob
$return.JobReference.JobId | Should Be $job.JobReference.JobId
$return.JobReference.ProjectId | Should Be $job.JobReference.ProjectId

ahandley commented:
Fixed the project issue.  Can't write a test for it today, but I'll put it in tomorrow with the Start- tests.

Also implemented the parameter sets.  Thanks for the hint on that.
